src/lib/Inference.hs

```diff
@@ -556,6 +518,33 @@ bindPat' (WithSrc pos pat) val = addSrcContext pos $ case pat of
     return $ env1 <> env2
   UPatVariant _ _ _ -> throw TypeErr "Variant not allowed in can't-fail pattern"
   UPatVariantLift _ _ -> throw TypeErr "Variant not allowed in can't-fail pattern"
+  UPatTable ps -> do
+    concreteIdxTy <- case getType val of
+      -- Table type is concrete (has no free variables).
+      TabTyAbs a | null (toList $ freeVars a) -> return $ absArgType a
+      -- Table type is not known to be concrete; assume it must be
+      -- Fin (length ps)
+      ty -> addContext context $ do
+          elemTy <- lift $ freshType TyKind
+          let idxTy = FixedIntRange 0 (fromIntegral $ length ps)
```

Comment:
I think that we actually have a `Fin` pattern declared in Syntax now!

Reply:
Looks like yes, we do, but it expects an Atom and not a haskell int. So it's between `FixedIntRange 0 (fromIntegral $ length ps)` and `Fin (IdxRepVal $ fromIntegral $ length ps)`. I'm tempted to leave as-is.

Reply:
Right, good point. It was a minor thing, we can leave it as is